### ISSUE TITLE
Erase Line (EL) control sequence not handled correctly

### ISSUE THREAD

**Issue by Pesa** [jenkinsci/ansicolor-plugin#151]
`gcc -fdiagnostics-color` [emits Erase Line control sequences](https://github.com/gcc-mirror/gcc/blob/master/gcc/color-macros.h#L24) in addition to the SGR strings for colorization. The ANSI color plugin doesn't seem to handle these control sequences correctly, thus rendering the colorized output unreadable.

### Test environment

- Jenkins Version: 2.163
- AnsiColor Version: 0.6.2
- Hostsystem: Ubuntu Linux

### Expected behavior

Readable warning/error messages from gcc in "Console output".

### Actual behavior

The following mess in "Console output". Note the extraneous `[K` sequences and in some cases a missing character just before the sequence (e.g. it should be "In member function...", but the last "n" is missing).

```
[K../tools/dump/ndndump.c[K In member functio[Kbool ndn::dump::NdnDump::printTcp(ndn::dump::OutputFormatter&, const uint8_t*, std::size_t) co[K’:
[K../tools/dump/ndndump.cpp:457:[Kerro[Kconst struct tcp[K’ has no member name[Kth_[K’; did you mea[Kd[K’?
   size_t tcpHdrLen = th->[Kth_[K * 4;
                          [K^~~[K
                          [Kd[K
[K../tools/dump/ndndump.c[K In member functio[Kbool ndn::dump::NdnDump::printUdp(ndn::dump::OutputFormatter&, const uint8_t*, std::size_t) co[K’:
[K../tools/dump/ndndump.cpp:486:[Kerro[Kconst struct udp[K’ has no member name[Kuh_u[K’
   size_t udpLen = endian::big_to_native(uh->[Kuh_u[K);
                                             [K^~~~[K
```

### Steps to reproduce the behavior

Just run `gcc -fdiagnostics-color` on anything that produces a compilation warning or error.

**Comment by Pesa** [comment]
The same snippet of log file viewed as "plain text":

```
[01m[K../tools/dump/ndndump.cpp:[m[K In member function ‘[01m[Kbool ndn::dump::NdnDump::printTcp(ndn::dump::OutputFormatter&, const uint8_t*, std::size_t) const[m[K’:
[01m[K../tools/dump/ndndump.cpp:457:26:[m[K [01;31m[Kerror: [m[K‘[01m[Kconst struct tcphdr[m[K’ has no member named ‘[01m[Kth_off[m[K’; did you mean ‘[01m[Kdoff[m[K’?
   size_t tcpHdrLen = th->[01;31m[Kth_off[m[K * 4;
                          [01;31m[K^~~~~~[m[K
                          [32m[Kdoff[m[K
[01m[K../tools/dump/ndndump.cpp:[m[K In member function ‘[01m[Kbool ndn::dump::NdnDump::printUdp(ndn::dump::OutputFormatter&, const uint8_t*, std::size_t) const[m[K’:
[01m[K../tools/dump/ndndump.cpp:486:45:[m[K [01;31m[Kerror: [m[K‘[01m[Kconst struct udphdr[m[K’ has no member named ‘[01m[Kuh_ulen[m[K’
   size_t udpLen = endian::big_to_native(uh->[01;31m[Kuh_ulen[m[K);
                                             [01;31m[K^~~~~~~[m[K
```

**Comment by dwnusbaum** [comment]
@Pesa Thanks for the report and the log snippet! Do you know if this is a regression from 0.5.x, or has it never worked?

**Comment by Pesa** [comment]
It's definitely a regression. Can't say exactly when it started happening, but it's been at least 2 or 3 months.

**Comment by Pesa** [comment]
Any updates on this?

**Comment by dwnusbaum** [comment]
I haven't had any time to investigate further. 

My guess is that the problem is that as of 0.6.x, the implementation of the coloring for Pipeline jobs has switched to using a different set of Jenkins APIs (`ConsoleAnnotator`+`MarkupText`), where the plugin adds markup to the rendered log to handle ANSI escape sequences, whereas previously the plugin used the `AnsiHtmlOutputStream` to directly write the log handling ANSI escape sequences while writing. The old implementation filtered all ANSI escape sequences as it wrote the logs, but the APIs used by the new logs do not allow content to be deleted (you can only add markup), so it has to [add markup to comment out the actual ANSI sequences themselves with `<!-- -->`](https://github.com/jenkinsci/ansicolor-plugin/blob/ff2264aa5d96ab819f96b904fec1766d529d194b/src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java#L100) when other markup is added. The new method relies on `AnsiHtmlOutputStream` calling `AnsiAttributeElement.emitHtml(String)` to signify that some markup should be emitted, but for invisible sequences like Erase Line, `AnsiHtmlOutputStream` never actually did anything in the past, it just filtered out the ANSI sequence, so `AnsiAttributeElement.emitHtml(String)` is not called and the new implementation is clueless that it needs to hide anything. 

#147 fixed a similar problem for text containing multiple Reset Graphics (SGR0) sequences in a row - the old implementation filtered the redundant sequences, but the new implementation only handled the sequences that actually resulted in emitting HTML to close open tags (usually the first one, but not always, and definitely not any of the subsequent ones).

If that is the problem, then my approach for a fix would be to make `AnsiHtmlOutputStream` override [`AnsiOutputStream.processEscapeLine`](https://github.com/jenkinsci/ansicolor-plugin/blob/ff2264aa5d96ab819f96b904fec1766d529d194b/src/main/java/hudson/plugins/ansicolor/AnsiOutputStream.java#L448) and all of the other `AnsiOutputStream.processX` methods that are not already implemented in `AnsiHtmlOutputStream`. The implementations should call [`AnsiAttributeElement.emitRedundantReset()`](https://github.com/jenkinsci/ansicolor-plugin/blob/ff2264aa5d96ab819f96b904fec1766d529d194b/src/main/java/hudson/plugins/ansicolor/AnsiAttributeElement.java#L36]) or a similar method (probably makes sense to rename the method to `emitInvisibleAnsiSequence`) so that `ColorConsoleAnnotator` runs through the logic to emit comments to hide AnsiSequences.

If someone has a lot of time to investigate, I would consider attempting to modify the current implementation to be a hybrid of the old and new approaches: Use a `ConsoleLogFilter` to hide the ANSI sequences and use `ConsoleAnnotator` to emit markup for the things that actual require markup. That might be impossible because `ConsoleLogFilters` run before `ConsoleAnnotators`, but I'm not sure without digging through code in Jenkins cores.

**Comment by dblock** [comment]
@Pesa Can you try to write a failing test/spec for this maybe?

**Comment by mathbunnyru** [comment]
Unfortunately, this plugin is not compatible with C++ projects :(

This is the simplest way to reproduce the problem
```
pipeline {
    options {
        ansiColor('xterm')
    }
    stages {
        stage('Incorrect gcc build') {
            steps {
                sh "g++ -fdiagnostics-color=always"
            }
        }
    }
}
```

`fatal error:` should be red.

**Comment by dwnusbaum** [comment]
I just released 0.6.3, most notably including #168 which should fix this issue and maybe some other issues that were lacking enough info to reproduce, and #171, which should fix #136.
